test/numbers.jl fails on 32-bit platform

```
$ make test-numbers
    JULIA test/numbers
     * numbers             exception on 1: ERROR: LoadError: test failed: ("0.3" != "0.3")
 in expression: repr(0.1 + 0.2) != "0.3"
 in error at error.jl:19
 in default_handler at test.jl:27
 in do_test at test.jl:50
 in runtests at /usr/local/src/julia/julia/test/testdefs.jl:66
 in anonymous at multi.jl:642
 in run_work_thunk at multi.jl:603
 in remotecall_fetch at multi.jl:676
 in remotecall_fetch at multi.jl:691
 in anonymous at task.jl:1614
while loading numbers.jl, in expression starting on line 318
ERROR: LoadError: LoadError: test failed: ("0.3" != "0.3")
 in expression: repr(0.1 + 0.2) != "0.3"
 in error at error.jl:19
 in default_handler at test.jl:27
 in do_test at test.jl:50
 in runtests at /usr/local/src/julia/julia/test/testdefs.jl:66
 in anonymous at multi.jl:642
 in run_work_thunk at multi.jl:603
 in remotecall_fetch at multi.jl:676
 in remotecall_fetch at multi.jl:691
 in anonymous at task.jl:1614
while loading numbers.jl, in expression starting on line 318
while loading /usr/local/src/julia/julia/test/runtests.jl, in expression starting on line 42

make[1]: *** [numbers] Error 1
make: *** [test-numbers] Error 2
[Rick@steelers julia]$ ./julia -e 'versioninfo()'
Julia Version 0.4.0-dev+2824
Commit 6bc53e6* (2015-01-19 20:14 UTC)
Platform Info:
  System: Linux (i686-redhat-linux)
  CPU: Genuine Intel(R) CPU           T2250  @ 1.73GHz
  WORD_SIZE: 32
  BLAS: libopenblas (DYNAMIC_ARCH NO_AFFINITY Banias)
  LAPACK: libopenblas
  LIBM: libopenlibm
  LLVM: libLLVM-3.3
```
Interestingly ...
```
$ ../julia
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "help()" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.4.0-dev+2824 (2015-01-19 20:14 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit 6bc53e6* (0 days old master)
|__/                   |  i686-redhat-linux

julia> repr(0.1 + 0.2)
"0.30000000000000004"

julia> include("testdefs.jl")
2-element Array{Union(UTF8String,ASCIIString),1}:
 "/usr/local/src/julia/julia/usr/local/share/julia/site/v0.4"
 "/usr/local/src/julia/julia/usr/share/julia/site/v0.4"      

julia> repr(0.1 + 0.2)
"0.30000000000000004"

julia> include("numbers.jl")
ERROR: LoadError: test failed: ("0.3" != "0.3")
 in expression: repr(0.1 + 0.2) != "0.3"
 in error at error.jl:19
 in default_handler at test.jl:27
 in do_test at test.jl:50
 in include at ./boot.jl:249
 in include_from_node1 at ./loading.jl:128
while loading /usr/local/src/julia/julia/test/numbers.jl, in expression starting on line 318

julia> repr(0.1 + 0.2)
"0.3"

```

--- a/base/float.jl
+++ b/base/float.jl
@@ -198,8 +198,6 @@ widen(::Type{Float32}) = Float64
 /(x::Float32, y::Float32) = box(Float32,div_float(unbox(Float32,x),unbox(Float32,y)))
 /(x::Float64, y::Float64) = box(Float64,div_float(unbox(Float64,x),unbox(Float64,y)))
 
-fma(x::Float32, y::Float32, z::Float32) = box(Float32,fma_float(unbox(Float32,x),unbox(Float32,y),unbox(Float32,z)))
-fma(x::Float64, y::Float64, z::Float64) = box(Float64,fma_float(unbox(Float64,x),unbox(Float64,y),unbox(Float64,z)))
 muladd(x::Float32, y::Float32, z::Float32) = box(Float32,muladd_float(unbox(Float32,x),unbox(Float32,y),unbox(Float32,z)))
 muladd(x::Float64, y::Float64, z::Float64) = box(Float64,muladd_float(unbox(Float64,x),unbox(Float64,y),unbox(Float64,z)))
 
@@ -385,6 +383,31 @@ end
     eps(::Type{Float32}) = $(box(Float32,unbox(UInt32,0x34000000)))
     eps(::Type{Float64}) = $(box(Float64,unbox(UInt64,0x3cb0000000000000)))
     eps() = eps(Float64)
+end
+
+# fused multiply-add
+fma_libm(x::Float32, y::Float32, z::Float32) =
+    ccall(("fmaf", libm_name), Float32, (Float32,Float32,Float32), x, y, z)
+fma_libm(x::Float64, y::Float64, z::Float64) =
+    ccall(("fma", libm_name), Float64, (Float64,Float64,Float64), x, y, z)
+fma_llvm(x::Float32, y::Float32, z::Float32) =
+    box(Float32,fma_float(unbox(Float32,x),unbox(Float32,y),unbox(Float32,z)))
+fma_llvm(x::Float64, y::Float64, z::Float64) =
+    box(Float64,fma_float(unbox(Float64,x),unbox(Float64,y),unbox(Float64,z)))
+# Disable LLVM's fma if it is incorrect, e.g. because LLVM falls back
+# onto a broken system libm; if so, use openlibm's fma instead
+# 1.0000305f0 = 1 + 1/2^15
+if fma_llvm(1.0000305f0, 1.0000305f0, -1.0f0) == 6.103609f-5
+    fma(x::Float32, y::Float32, z::Float32) = fma_llvm(x,y,z)
+else
+    fma(x::Float32, y::Float32, z::Float32) = fma_libm(x,y,z)
+end
+# 1.0000000009313226 = 1 + 1/2^30
+if (fma_llvm(1.0000000009313226, 1.0000000009313226, -1.0) ==
+    1.8626451500983188e-9)
+    fma(x::Float64, y::Float64, z::Float64) = fma_llvm(x,y,z)
+else
+    fma(x::Float64, y::Float64, z::Float64) = fma_libm(x,y,z)
 end
 
 ## byte order swaps for arbitrary-endianness serialization/deserialization ##

